Add carrier current location

melissa-hong opened the pull request:
<!--Read comments, before commiting pull request read checklist again

# Checklist:

- I have performed a self-review of my own code,
- I have commented my code, particularly in hard-to-understand areas,
- I have made corresponding changes to the documentation,
- I have added tests that prove my fix is effective or that my feature works,
- New and existing unit tests pass locally with my changes ("bundle exec rake"),
- Title include "WIP" if work is in progress.

-->

Resolves #108

### Description
Add a current location to a carrier to distinguish between home location.

Rename existing `location_id` to `home_location_id` and add a new `current_location_id`. Populate  current location with home location initially

### Type of change
* New feature (non-breaking change which adds functionality)

### How Has This Been Tested?
- Updated test for carrier model.
- Updated carrier feature test.

### Screenshots
`/carriers/new`, `/carriers/:id/edit`
![image](https://user-images.githubusercontent.com/478300/66727752-06015880-edf6-11e9-8ac7-fcb472e2739a.png)

`/carriers/:id`
![image](https://user-images.githubusercontent.com/478300/66803179-b98c4a80-eed3-11e9-83b9-b21ece951b5e.png)

Commit:
Add dropdown to carrier view to set/update current location

- Replace references to carrier.location to home_location

Commit:
Update tests for current location on carrier

DamirSvrtan commented:
Hey @melissa-hong, thanks a lot for your contribution! Could I please ask you to merge master into your own branch so that the specs can pass? We unfortunately had a master that had a failing spec and now all the PR's are showing failing Travis CI builds.

melissa-hong commented:
@pollygee @abronkema this is ready for review now. Thanks!

dixpac commented on app/models/carrier.rb:
```diff
@@ -2,6 +2,8 @@
 
 class Carrier < ApplicationRecord
   belongs_to :location
+  alias_attribute :home_location, :location
```
Nice trick with alias here, but I think it would be better to rename location column -> home_location at the moment. Project is still in development phase we don't have any production customers so no need to be afraid of renaming the column :)

pollygee replied:
Good point @dixpac I think in the long run having a name that is not ambiguous can save us pain in the future when working on this and asking ourselves "is location home_location or current_location".

melissa-hong replied:
Ah I see - makes sense. I've updated the migration now (realised after I could have used a `rename_column` but kept as is since adding as a reference adds it as a bigint instead of integer)

dixpac commented on lib/tasks/carriers.rake:
```diff
@@ -0,0 +1,8 @@
+namespace :carriers do
```
Same as above, awesome idea if we had production customers, but since we don't I think it would be easier to stick this `backfilling` inside the migration itself

dixpac commented on db/migrate/20191014030158_add_current_location_to_carrier.rb:
```diff
@@ -0,0 +1,5 @@
+class AddCurrentLocationToCarrier < ActiveRecord::Migration[5.2]
+  def change
+    add_reference :carriers, :current_location, foreign_key: { to_table: :locations }
```
👍

dixpac commented on app/models/carrier.rb:
```diff
@@ -2,6 +2,8 @@
 class Carrier < ApplicationRecord
   belongs_to :location
+  alias_attribute :home_location, :location
+  belongs_to :current_location, class_name: 'Location', optional: true
```
@abronkema is it possible to have `Carrier` without the `current_location`?

dixpac commented on spec/fixtures/locations.yml:
```diff
@@ -1,2 +1,4 @@
-location:
```
🥇 👏

dixpac reviewed (commented):
@melissa-hong awesome work! I've just left few small comments

Commit:
Add Home and Current location to Carriers

- Populate home and current location with existing location_id
- Remove ambiguous location_id
- Make home and current location required

Commit:
Update seeds file to populate current location on carrier

- And default to home location

melissa-hong commented:
Thanks @dixpac! I've addressed all your comments now, the spec failures seem to be due to the flaky carrier specs addressed in #140

pollygee commented:
Looks great @melissa-hong!  There were some changes to the features, can you merge master into your branch and fix the conflicts please?  Thanks so much!

melissa-hong commented:
@pollygee have merged master in now!

dixpac commented:
Thanks @melissa-hong you did awesome an work on this

dixpac commented on db/migrate/20191016050734_add_home_and_current_location_to_carrier.rb:
```diff
@@ -0,0 +1,12 @@
+    add_reference :carriers, :current_location, foreign_key: { to_table: :locations }
+
+    Carrier.all.each do |carrier|
+      carrier.update(home_location_id: carrier.location_id, current_location_id: carrier.location_id)
```
@melissa-hong  just tinny little tip for the future(not super important for this case)...always make sure that migrations are reversible, this here will fail on `bin/rails db:rollback` :)

melissa-hong replied:
oh that's a good point - sorry about that. Thanks for the review!
